{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "babd3ee0",
   "metadata": {},
   "source": [
    "This script is modified from `notebooks/monarch_gnn/3_create_dataset_pair_tensor_nd_validate.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "710d936d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/src/modspy_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3173e30c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import torch\n",
    "import optuna\n",
    "from comet_ml import Experiment\n",
    "from pytorch_lightning.loggers import CometLogger\n",
    "from torch_geometric.transforms import AddSelfLoops\n",
    "import wandb\n",
    "\n",
    "\n",
    "import dask\n",
    "import dask.dataframe as dd\n",
    "\n",
    "from models.embed import MetaPath2VecLightningModule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f528f5c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "Total nodes: 862115\n",
      "Total node types: 88\n",
      "\n",
      "Total edges: 11412471\n",
      "Total edge types: 289                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "monarch = torch.load('./data/05_model_input/2024-02-monarch_heterodata_v1.pt', map_location='cpu')\n",
    "\n",
    "print(\n",
    "    f\"\"\" \n",
    "Total nodes: {monarch.num_nodes}\n",
    "Total node types: {len(monarch.node_types)}\n",
    "\n",
    "Total edges: {monarch.num_edges}\n",
    "Total edge types: {len(monarch.edge_types)}                \n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "# # Adding self loops to avoid 1. nodes without any edge, 2. consider intragenic modifier\n",
    "# transform = AddSelfLoops()\n",
    "# monarch = transform(monarch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "71a65b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_df = pd.read_csv('./data/04_feature/2024-06-11-modifier_merged_dataset.tsv', sep=\"\\t\")\n",
    "# dataset_df = catalog.load(\"modifiers\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "abfe5a08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>datasource</th>\n",
       "      <th>species</th>\n",
       "      <th>modifier_gene_symbol</th>\n",
       "      <th>target_gene_symbol</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>go_n_common_ancestors_max</th>\n",
       "      <th>go_n_common_ancestors_avg</th>\n",
       "      <th>go_n_common_ancestors_bma</th>\n",
       "      <th>go_n_union_ancestors_max</th>\n",
       "      <th>go_n_union_ancestors_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>homology</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>coexpression_transferred</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>ATR</td>\n",
       "      <td>1.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.960317</td>\n",
       "      <td>6.109155</td>\n",
       "      <td>101.0</td>\n",
       "      <td>28.846782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DCLRE1C</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>4.163090</td>\n",
       "      <td>7.260417</td>\n",
       "      <td>75.0</td>\n",
       "      <td>29.659375</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>292.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>540.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>321.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>779.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SMARCAL1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>5.598765</td>\n",
       "      <td>10.268041</td>\n",
       "      <td>87.0</td>\n",
       "      <td>31.152206</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>506.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>602.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TRIP13</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>4.515110</td>\n",
       "      <td>11.198020</td>\n",
       "      <td>128.0</td>\n",
       "      <td>39.033929</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>505.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MUS81</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>4.599469</td>\n",
       "      <td>9.086957</td>\n",
       "      <td>88.0</td>\n",
       "      <td>28.369439</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>768.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>919.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 73 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  datasource       species modifier_gene_symbol target_gene_symbol  \\\n",
       "0        JVL  Homo sapiens                 ETV1                ATR   \n",
       "1        JVL  Homo sapiens              DCLRE1C              BRCA1   \n",
       "2        JVL  Homo sapiens             SMARCAL1              BRCA1   \n",
       "3        JVL  Homo sapiens               TRIP13              BRCA1   \n",
       "4        JVL  Homo sapiens                MUS81              BRCA2   \n",
       "\n",
       "   is_modifier  go_n_common_ancestors_max  go_n_common_ancestors_avg  \\\n",
       "0          1.0                       59.0                   2.960317   \n",
       "1          1.0                       22.0                   4.163090   \n",
       "2          1.0                       45.0                   5.598765   \n",
       "3          1.0                       43.0                   4.515110   \n",
       "4          1.0                       29.0                   4.599469   \n",
       "\n",
       "   go_n_common_ancestors_bma  go_n_union_ancestors_max  \\\n",
       "0                   6.109155                     101.0   \n",
       "1                   7.260417                      75.0   \n",
       "2                  10.268041                      87.0   \n",
       "3                  11.198020                     128.0   \n",
       "4                   9.086957                      88.0   \n",
       "\n",
       "   go_n_union_ancestors_avg  ...  homology  coexpression  \\\n",
       "0                 28.846782  ...       0.0           0.0   \n",
       "1                 29.659375  ...       0.0          80.0   \n",
       "2                 31.152206  ...       0.0          49.0   \n",
       "3                 39.033929  ...       0.0         243.0   \n",
       "4                 28.369439  ...       0.0           0.0   \n",
       "\n",
       "   coexpression_transferred  experiments  experiments_transferred  database  \\\n",
       "0                       0.0          0.0                      0.0       0.0   \n",
       "1                       0.0        292.0                      0.0     540.0   \n",
       "2                      62.0          0.0                      0.0       0.0   \n",
       "3                     110.0          0.0                      0.0       0.0   \n",
       "4                      61.0          0.0                      0.0     500.0   \n",
       "\n",
       "   database_transferred  textmining  textmining_transferred  combined_score  \n",
       "0                   0.0         0.0                     0.0             0.0  \n",
       "1                   0.0       321.0                    83.0           779.0  \n",
       "2                   0.0       506.0                   204.0           602.0  \n",
       "3                   0.0       285.0                    95.0           505.0  \n",
       "4                   0.0       768.0                   350.0           919.0  \n",
       "\n",
       "[5 rows x 73 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_df['species'] = 'Homo sapiens'\n",
    "dataset_df.loc[dataset_df['datasource']=='MTG', 'species'] = 'Caenorhabditis elegans'\n",
    "\n",
    "# Reordering important columns\n",
    "desired_first_columns = ['datasource', 'species', 'modifier_gene_symbol',\n",
    "                         'target_gene_symbol', 'is_modifier'] \n",
    "remaining_columns = [col for col in dataset_df.columns if col not in desired_first_columns]\n",
    "new_column_order = desired_first_columns + remaining_columns\n",
    "dataset_df = dataset_df[new_column_order]\n",
    "dataset_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee888365",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_df = dd.read_parquet('./data/02_intermediate/monarch/nodes_with_type_idx')  \n",
    "edges_df = dd.read_parquet('./data/02_intermediate/monarch/edges_pre_df_reduction_v2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fce840d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "862115\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>name</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>in_taxon_label</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>APO:0000017</th>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000001</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>entity</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000002</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>continuant</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000003</th>\n",
       "      <td>biolink:BiologicalProcessOrActivity</td>\n",
       "      <td>occurrent</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000004</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>independent continuant</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        category                    name  \\\n",
       "id                                                                         \n",
       "APO:0000017            biolink:PhenotypicFeature                    <NA>   \n",
       "BFO:0000001                   biolink:NamedThing                  entity   \n",
       "BFO:0000002                   biolink:NamedThing              continuant   \n",
       "BFO:0000003  biolink:BiologicalProcessOrActivity               occurrent   \n",
       "BFO:0000004                   biolink:NamedThing  independent continuant   \n",
       "\n",
       "            in_taxon in_taxon_label symbol  type_index  \n",
       "id                                                      \n",
       "APO:0000017     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000001     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000002     <NA>           <NA>   <NA>           1  \n",
       "BFO:0000003     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000004     <NA>           <NA>   <NA>           2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(nodes_df.index.shape[0].compute())\n",
    "display(nodes_df.head())\n",
    "\n",
    "# Use dataset's source and target gene id to pull the `type_index` column From `nodes_df``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6138482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>name</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>in_taxon_label</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HGNC:613</th>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>APOE</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>516790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              category  name        in_taxon in_taxon_label symbol  type_index\n",
       "id                                                                            \n",
       "HGNC:613  biolink:Gene  APOE  NCBITaxon:9606   Homo sapiens   APOE      516790"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_df[(nodes_df['category']=='biolink:Gene') & (nodes_df['symbol']=='APOE') & (nodes_df['in_taxon_label']=='Homo sapiens')].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1cf1356e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                              <NA>\n",
       "1                      Homo sapiens\n",
       "2                      Mus musculus\n",
       "3                     Gallus gallus\n",
       "4                        Bos taurus\n",
       "5                 Rattus norvegicus\n",
       "0           Drosophila melanogaster\n",
       "1                        Sus scrofa\n",
       "2            Canis lupus familiaris\n",
       "3          Dictyostelium discoideum\n",
       "4         Schizosaccharomyces pombe\n",
       "5    Saccharomyces cerevisiae S288C\n",
       "6            Caenorhabditis elegans\n",
       "7                    Xenopus laevis\n",
       "8                Xenopus tropicalis\n",
       "9                       Danio rerio\n",
       "Name: in_taxon_label, dtype: string"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_df['in_taxon_label'].unique().compute()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c57d3f",
   "metadata": {},
   "source": [
    "### Merging Node Index from Monarch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c9f7fca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>datasource</th>\n",
       "      <th>species</th>\n",
       "      <th>modifier_gene_symbol</th>\n",
       "      <th>target_gene_symbol</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>go_n_common_ancestors_max</th>\n",
       "      <th>go_n_common_ancestors_avg</th>\n",
       "      <th>go_n_common_ancestors_bma</th>\n",
       "      <th>go_n_union_ancestors_max</th>\n",
       "      <th>go_n_union_ancestors_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>ATR</td>\n",
       "      <td>1.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>2.960317</td>\n",
       "      <td>6.109155</td>\n",
       "      <td>101.0</td>\n",
       "      <td>28.846782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>517533</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>523802.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>DCLRE1C</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>4.163090</td>\n",
       "      <td>7.260417</td>\n",
       "      <td>75.0</td>\n",
       "      <td>29.659375</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>540.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>321.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>779.0</td>\n",
       "      <td>518104</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>521725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SMARCAL1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>5.598765</td>\n",
       "      <td>10.268041</td>\n",
       "      <td>87.0</td>\n",
       "      <td>31.152206</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>506.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>602.0</td>\n",
       "      <td>518104</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>551530.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TRIP13</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>4.515110</td>\n",
       "      <td>11.198020</td>\n",
       "      <td>128.0</td>\n",
       "      <td>39.033929</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>505.0</td>\n",
       "      <td>518104</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>555869.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>MUS81</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>4.599469</td>\n",
       "      <td>9.086957</td>\n",
       "      <td>88.0</td>\n",
       "      <td>28.369439</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>768.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>919.0</td>\n",
       "      <td>518106</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>537363.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5791</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>egl-13</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.379699</td>\n",
       "      <td>4.038462</td>\n",
       "      <td>162.0</td>\n",
       "      <td>31.414583</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>82814.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5792</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>F41C6.7</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.646259</td>\n",
       "      <td>1.851064</td>\n",
       "      <td>112.0</td>\n",
       "      <td>21.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>95948.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5793</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>F48E3.8</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.589286</td>\n",
       "      <td>2.121212</td>\n",
       "      <td>110.0</td>\n",
       "      <td>21.241667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5794</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>C39D10.7</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.653846</td>\n",
       "      <td>1.947368</td>\n",
       "      <td>103.0</td>\n",
       "      <td>17.625000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>94433.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5795</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>tmc-2</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.699029</td>\n",
       "      <td>1.826087</td>\n",
       "      <td>115.0</td>\n",
       "      <td>21.920833</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>93262.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5796 rows Ã— 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     datasource                 species modifier_gene_symbol  \\\n",
       "0           JVL            Homo sapiens                 ETV1   \n",
       "1           JVL            Homo sapiens              DCLRE1C   \n",
       "2           JVL            Homo sapiens             SMARCAL1   \n",
       "3           JVL            Homo sapiens               TRIP13   \n",
       "4           JVL            Homo sapiens                MUS81   \n",
       "...         ...                     ...                  ...   \n",
       "5791        MTG  Caenorhabditis elegans               egl-13   \n",
       "5792        MTG  Caenorhabditis elegans              F41C6.7   \n",
       "5793        MTG  Caenorhabditis elegans              F48E3.8   \n",
       "5794        MTG  Caenorhabditis elegans             C39D10.7   \n",
       "5795        MTG  Caenorhabditis elegans                tmc-2   \n",
       "\n",
       "     target_gene_symbol  is_modifier  go_n_common_ancestors_max  \\\n",
       "0                   ATR          1.0                       59.0   \n",
       "1                 BRCA1          1.0                       22.0   \n",
       "2                 BRCA1          1.0                       45.0   \n",
       "3                 BRCA1          1.0                       43.0   \n",
       "4                 BRCA2          1.0                       29.0   \n",
       "...                 ...          ...                        ...   \n",
       "5791              zyg-1          0.0                        8.0   \n",
       "5792              zyg-1          0.0                        3.0   \n",
       "5793              zyg-1          0.0                        5.0   \n",
       "5794              zyg-1          0.0                        3.0   \n",
       "5795              zyg-1          0.0                        3.0   \n",
       "\n",
       "      go_n_common_ancestors_avg  go_n_common_ancestors_bma  \\\n",
       "0                      2.960317                   6.109155   \n",
       "1                      4.163090                   7.260417   \n",
       "2                      5.598765                  10.268041   \n",
       "3                      4.515110                  11.198020   \n",
       "4                      4.599469                   9.086957   \n",
       "...                         ...                        ...   \n",
       "5791                   2.379699                   4.038462   \n",
       "5792                   1.646259                   1.851064   \n",
       "5793                   1.589286                   2.121212   \n",
       "5794                   1.653846                   1.947368   \n",
       "5795                   1.699029                   1.826087   \n",
       "\n",
       "      go_n_union_ancestors_max  go_n_union_ancestors_avg  ...  \\\n",
       "0                        101.0                 28.846782  ...   \n",
       "1                         75.0                 29.659375  ...   \n",
       "2                         87.0                 31.152206  ...   \n",
       "3                        128.0                 39.033929  ...   \n",
       "4                         88.0                 28.369439  ...   \n",
       "...                        ...                       ...  ...   \n",
       "5791                     162.0                 31.414583  ...   \n",
       "5792                     112.0                 21.500000  ...   \n",
       "5793                     110.0                 21.241667  ...   \n",
       "5794                     103.0                 17.625000  ...   \n",
       "5795                     115.0                 21.920833  ...   \n",
       "\n",
       "      experiments_transferred  database  database_transferred  textmining  \\\n",
       "0                         0.0       0.0                   0.0         0.0   \n",
       "1                         0.0     540.0                   0.0       321.0   \n",
       "2                         0.0       0.0                   0.0       506.0   \n",
       "3                         0.0       0.0                   0.0       285.0   \n",
       "4                         0.0     500.0                   0.0       768.0   \n",
       "...                       ...       ...                   ...         ...   \n",
       "5791                      0.0       0.0                   0.0         0.0   \n",
       "5792                      0.0       0.0                   0.0         0.0   \n",
       "5793                      0.0       0.0                   0.0         0.0   \n",
       "5794                      0.0       0.0                   0.0         0.0   \n",
       "5795                      0.0       0.0                   0.0         0.0   \n",
       "\n",
       "      textmining_transferred  combined_score  target_type_index      category  \\\n",
       "0                        0.0             0.0             517533  biolink:Gene   \n",
       "1                       83.0           779.0             518104  biolink:Gene   \n",
       "2                      204.0           602.0             518104  biolink:Gene   \n",
       "3                       95.0           505.0             518104  biolink:Gene   \n",
       "4                      350.0           919.0             518106  biolink:Gene   \n",
       "...                      ...             ...                ...           ...   \n",
       "5791                     0.0             0.0              86692  biolink:Gene   \n",
       "5792                     0.0             0.0              86692  biolink:Gene   \n",
       "5793                     0.0             0.0              86692          <NA>   \n",
       "5794                     0.0             0.0              86692  biolink:Gene   \n",
       "5795                     0.0             0.0              86692  biolink:Gene   \n",
       "\n",
       "            in_taxon  modifier_type_index  \n",
       "0     NCBITaxon:9606             523802.0  \n",
       "1     NCBITaxon:9606             521725.0  \n",
       "2     NCBITaxon:9606             551530.0  \n",
       "3     NCBITaxon:9606             555869.0  \n",
       "4     NCBITaxon:9606             537363.0  \n",
       "...              ...                  ...  \n",
       "5791  NCBITaxon:6239              82814.0  \n",
       "5792  NCBITaxon:6239              95948.0  \n",
       "5793            <NA>                  NaN  \n",
       "5794  NCBITaxon:6239              94433.0  \n",
       "5795  NCBITaxon:6239              93262.0  \n",
       "\n",
       "[5796 rows x 77 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_pdf = nodes_df.compute()\n",
    "_dataset = dataset_df.merge(nodes_pdf, how='left', left_on=['species','target_gene_symbol'], right_on=['in_taxon_label','symbol'])\n",
    "_dataset.rename(columns={'type_index': 'target_type_index'}, inplace=True)\n",
    "_dataset.drop(columns=['category','name','in_taxon_label','in_taxon','symbol'], inplace=True)\n",
    "dataset_w_ninfo = _dataset.merge(nodes_pdf, how='left', left_on=['species','modifier_gene_symbol'], right_on=['in_taxon_label','symbol'])\n",
    "dataset_w_ninfo.rename(columns={'type_index': 'modifier_type_index'}, inplace=True)\n",
    "dataset_w_ninfo.drop(columns=['name','in_taxon_label','symbol'], inplace=True)\n",
    "del(_dataset)\n",
    "del(nodes_pdf)\n",
    "dataset_w_ninfo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "324d864c",
   "metadata": {},
   "source": [
    "### Did we detect all IDs for the genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0d3ff36f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "print(dataset_w_ninfo['target_type_index'].unique().shape == dataset_w_ninfo['target_gene_symbol'].unique().shape)\n",
    "print(dataset_w_ninfo['modifier_type_index'].unique().shape == dataset_w_ninfo['modifier_gene_symbol'].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ce6be17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "254 genes are missing/not recognized from Monarch\n"
     ]
    }
   ],
   "source": [
    "print(f\"{dataset_w_ninfo['modifier_gene_symbol'].unique().shape[0] - dataset_w_ninfo['modifier_type_index'].unique().shape[0]} genes are missing/not recognized from Monarch\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e5093a1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unrecognized Modifiers:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>datasource</th>\n",
       "      <th>species</th>\n",
       "      <th>modifier_gene_symbol</th>\n",
       "      <th>target_gene_symbol</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>go_n_common_ancestors_max</th>\n",
       "      <th>go_n_common_ancestors_avg</th>\n",
       "      <th>go_n_common_ancestors_bma</th>\n",
       "      <th>go_n_union_ancestors_max</th>\n",
       "      <th>go_n_union_ancestors_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>3.151899</td>\n",
       "      <td>7.222222</td>\n",
       "      <td>125.0</td>\n",
       "      <td>27.442857</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2317</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>apc-17</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     datasource                 species modifier_gene_symbol  \\\n",
       "672         MTG  Caenorhabditis elegans                mat-2   \n",
       "2317        MTG  Caenorhabditis elegans               apc-17   \n",
       "\n",
       "     target_gene_symbol  is_modifier  go_n_common_ancestors_max  \\\n",
       "672               zyg-1          1.0                       34.0   \n",
       "2317              zyg-1          1.0                        0.0   \n",
       "\n",
       "      go_n_common_ancestors_avg  go_n_common_ancestors_bma  \\\n",
       "672                    3.151899                   7.222222   \n",
       "2317                   0.000000                   0.000000   \n",
       "\n",
       "      go_n_union_ancestors_max  go_n_union_ancestors_avg  ...  \\\n",
       "672                      125.0                 27.442857  ...   \n",
       "2317                       0.0                  0.000000  ...   \n",
       "\n",
       "      experiments_transferred  database  database_transferred  textmining  \\\n",
       "672                       0.0       0.0                   0.0         0.0   \n",
       "2317                      0.0       0.0                   0.0         0.0   \n",
       "\n",
       "      textmining_transferred  combined_score  target_type_index  category  \\\n",
       "672                      0.0             0.0              86692      <NA>   \n",
       "2317                     0.0             0.0              86692      <NA>   \n",
       "\n",
       "      in_taxon  modifier_type_index  \n",
       "672       <NA>                  NaN  \n",
       "2317      <NA>                  NaN  \n",
       "\n",
       "[2 rows x 77 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Unrecognized Modifiers:\")\n",
    "dataset_w_ninfo[(dataset_w_ninfo['modifier_type_index'].isna() & (dataset_w_ninfo['is_modifier']==1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8a6cc351",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing indecies\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>datasource</th>\n",
       "      <th>species</th>\n",
       "      <th>modifier_gene_symbol</th>\n",
       "      <th>target_gene_symbol</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>go_n_common_ancestors_max</th>\n",
       "      <th>go_n_common_ancestors_avg</th>\n",
       "      <th>go_n_common_ancestors_bma</th>\n",
       "      <th>go_n_union_ancestors_max</th>\n",
       "      <th>go_n_union_ancestors_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>experiments_transferred</th>\n",
       "      <th>database</th>\n",
       "      <th>database_transferred</th>\n",
       "      <th>textmining</th>\n",
       "      <th>textmining_transferred</th>\n",
       "      <th>combined_score</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>OLIDA</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NOT_HGNC 4q35</td>\n",
       "      <td>SMCHD1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>551559</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>623</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>T20B6.3</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Y51A2D.7</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.774194</td>\n",
       "      <td>3.031250</td>\n",
       "      <td>121.0</td>\n",
       "      <td>31.575000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>3.151899</td>\n",
       "      <td>7.222222</td>\n",
       "      <td>125.0</td>\n",
       "      <td>27.442857</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>690</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>Y71H2AM.13</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5661</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>W08E12.3</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.600000</td>\n",
       "      <td>1.727273</td>\n",
       "      <td>102.0</td>\n",
       "      <td>17.250000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5737</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>F09C3.2</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>2.393939</td>\n",
       "      <td>2.727273</td>\n",
       "      <td>117.0</td>\n",
       "      <td>23.912500</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5748</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>D1081.7</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5787</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>F59D12.1</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.769231</td>\n",
       "      <td>2.911111</td>\n",
       "      <td>106.0</td>\n",
       "      <td>22.370000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5793</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>F48E3.8</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.589286</td>\n",
       "      <td>2.121212</td>\n",
       "      <td>110.0</td>\n",
       "      <td>21.241667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>255 rows Ã— 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     datasource                 species modifier_gene_symbol  \\\n",
       "499       OLIDA            Homo sapiens        NOT_HGNC 4q35   \n",
       "623         MTG  Caenorhabditis elegans              T20B6.3   \n",
       "642         MTG  Caenorhabditis elegans             Y51A2D.7   \n",
       "672         MTG  Caenorhabditis elegans                mat-2   \n",
       "690         MTG  Caenorhabditis elegans           Y71H2AM.13   \n",
       "...         ...                     ...                  ...   \n",
       "5661        MTG  Caenorhabditis elegans             W08E12.3   \n",
       "5737        MTG  Caenorhabditis elegans              F09C3.2   \n",
       "5748        MTG  Caenorhabditis elegans              D1081.7   \n",
       "5787        MTG  Caenorhabditis elegans             F59D12.1   \n",
       "5793        MTG  Caenorhabditis elegans              F48E3.8   \n",
       "\n",
       "     target_gene_symbol  is_modifier  go_n_common_ancestors_max  \\\n",
       "499              SMCHD1          0.0                        0.0   \n",
       "623               zyg-1          0.0                        0.0   \n",
       "642               zyg-1          0.0                        8.0   \n",
       "672               zyg-1          1.0                       34.0   \n",
       "690               zyg-1          0.0                        0.0   \n",
       "...                 ...          ...                        ...   \n",
       "5661              zyg-1          0.0                        3.0   \n",
       "5737              zyg-1          0.0                        9.0   \n",
       "5748              zyg-1          0.0                        0.0   \n",
       "5787              zyg-1          0.0                       10.0   \n",
       "5793              zyg-1          0.0                        5.0   \n",
       "\n",
       "      go_n_common_ancestors_avg  go_n_common_ancestors_bma  \\\n",
       "499                    0.000000                   0.000000   \n",
       "623                    0.000000                   0.000000   \n",
       "642                    2.774194                   3.031250   \n",
       "672                    3.151899                   7.222222   \n",
       "690                    0.000000                   0.000000   \n",
       "...                         ...                        ...   \n",
       "5661                   1.600000                   1.727273   \n",
       "5737                   2.393939                   2.727273   \n",
       "5748                   0.000000                   0.000000   \n",
       "5787                   2.769231                   2.911111   \n",
       "5793                   1.589286                   2.121212   \n",
       "\n",
       "      go_n_union_ancestors_max  go_n_union_ancestors_avg  ...  \\\n",
       "499                        0.0                  0.000000  ...   \n",
       "623                        0.0                  0.000000  ...   \n",
       "642                      121.0                 31.575000  ...   \n",
       "672                      125.0                 27.442857  ...   \n",
       "690                        0.0                  0.000000  ...   \n",
       "...                        ...                       ...  ...   \n",
       "5661                     102.0                 17.250000  ...   \n",
       "5737                     117.0                 23.912500  ...   \n",
       "5748                       0.0                  0.000000  ...   \n",
       "5787                     106.0                 22.370000  ...   \n",
       "5793                     110.0                 21.241667  ...   \n",
       "\n",
       "      experiments_transferred  database  database_transferred  textmining  \\\n",
       "499                       0.0       0.0                   0.0         0.0   \n",
       "623                       0.0       0.0                   0.0         0.0   \n",
       "642                       0.0       0.0                   0.0         0.0   \n",
       "672                       0.0       0.0                   0.0         0.0   \n",
       "690                       0.0       0.0                   0.0         0.0   \n",
       "...                       ...       ...                   ...         ...   \n",
       "5661                      0.0       0.0                   0.0         0.0   \n",
       "5737                      0.0       0.0                   0.0         0.0   \n",
       "5748                      0.0       0.0                   0.0         0.0   \n",
       "5787                      0.0       0.0                   0.0         0.0   \n",
       "5793                      0.0       0.0                   0.0         0.0   \n",
       "\n",
       "      textmining_transferred  combined_score  target_type_index  category  \\\n",
       "499                      0.0             0.0             551559      <NA>   \n",
       "623                      0.0             0.0              86692      <NA>   \n",
       "642                      0.0             0.0              86692      <NA>   \n",
       "672                      0.0             0.0              86692      <NA>   \n",
       "690                      0.0             0.0              86692      <NA>   \n",
       "...                      ...             ...                ...       ...   \n",
       "5661                     0.0             0.0              86692      <NA>   \n",
       "5737                     0.0             0.0              86692      <NA>   \n",
       "5748                     0.0             0.0              86692      <NA>   \n",
       "5787                     0.0             0.0              86692      <NA>   \n",
       "5793                     0.0             0.0              86692      <NA>   \n",
       "\n",
       "      in_taxon  modifier_type_index  \n",
       "499       <NA>                  NaN  \n",
       "623       <NA>                  NaN  \n",
       "642       <NA>                  NaN  \n",
       "672       <NA>                  NaN  \n",
       "690       <NA>                  NaN  \n",
       "...        ...                  ...  \n",
       "5661      <NA>                  NaN  \n",
       "5737      <NA>                  NaN  \n",
       "5748      <NA>                  NaN  \n",
       "5787      <NA>                  NaN  \n",
       "5793      <NA>                  NaN  \n",
       "\n",
       "[255 rows x 77 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Missing indecies\")\n",
    "dataset_w_ninfo[(dataset_w_ninfo['modifier_type_index'].isna()) | (dataset_w_ninfo['target_type_index'].isna())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b02ae79a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Do we now have proper indecies?\n",
      "Target genes: True\n",
      "Modifier genes: True\n"
     ]
    }
   ],
   "source": [
    "# Dropping these NA rows\n",
    "dataset_w_ninfo.dropna(subset=['in_taxon','modifier_type_index'], inplace=True)\n",
    "\n",
    "print(f\"Do we now have proper indecies?\")\n",
    "print(f\"Target genes: {dataset_w_ninfo['target_type_index'].unique().shape == dataset_w_ninfo['target_gene_symbol'].unique().shape}\")\n",
    "print(f\"Modifier genes: {dataset_w_ninfo['modifier_type_index'].unique().shape == dataset_w_ninfo['modifier_gene_symbol'].unique().shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41567b04",
   "metadata": {},
   "source": [
    "### Adding target-modifier pairs as edges of Monarch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ed40cdf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>modifier_type_index</th>\n",
       "      <th>is_modifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>517533</td>\n",
       "      <td>523802</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>518104</td>\n",
       "      <td>521725</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>518104</td>\n",
       "      <td>551530</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>518104</td>\n",
       "      <td>555869</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>518106</td>\n",
       "      <td>537363</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5790</th>\n",
       "      <td>86692</td>\n",
       "      <td>86202</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5791</th>\n",
       "      <td>86692</td>\n",
       "      <td>82814</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5792</th>\n",
       "      <td>86692</td>\n",
       "      <td>95948</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5794</th>\n",
       "      <td>86692</td>\n",
       "      <td>94433</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5795</th>\n",
       "      <td>86692</td>\n",
       "      <td>93262</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5541 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      target_type_index  modifier_type_index  is_modifier\n",
       "0                517533               523802          1.0\n",
       "1                518104               521725          1.0\n",
       "2                518104               551530          1.0\n",
       "3                518104               555869          1.0\n",
       "4                518106               537363          1.0\n",
       "...                 ...                  ...          ...\n",
       "5790              86692                86202          0.0\n",
       "5791              86692                82814          0.0\n",
       "5792              86692                95948          0.0\n",
       "5794              86692                94433          0.0\n",
       "5795              86692                93262          0.0\n",
       "\n",
       "[5541 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_w_ninfo['modifier_type_index'] = dataset_w_ninfo['modifier_type_index'].astype(int)\n",
    "dataset_w_ninfo[['target_type_index','modifier_type_index','is_modifier']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b89a3097",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset_w_ninfo.to_csv('./data/04_feature/2024-06-12-merged-dataset-nodeinfo.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "d92496ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([[517533, 523802,      1],\n",
      "        [518104, 521725,      1],\n",
      "        [518104, 551530,      1],\n",
      "        ...,\n",
      "        [ 86692,  95948,      0],\n",
      "        [ 86692,  94433,      0],\n",
      "        [ 86692,  93262,      0]], dtype=torch.int32)\n",
      "torch.Size([5541, 3])\n"
     ]
    }
   ],
   "source": [
    "# Convert DataFrame to tensor\n",
    "dataset_arr = dataset_w_ninfo[['target_type_index','modifier_type_index','is_modifier']].to_numpy()\n",
    "data_tensor = torch.from_numpy(dataset_arr).to(torch.int)\n",
    "print(data_tensor)\n",
    "print(data_tensor.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "25070552",
   "metadata": {},
   "outputs": [],
   "source": [
    "# torch.save(data_t, './data/05_model_input/2024-06-12-merged-dataset-nodeinfo.pt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a3592b1",
   "metadata": {},
   "source": [
    "<b>Get GNN Model<b> [Need GPU]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61b513d8",
   "metadata": {},
   "source": [
    "Loading pre-trained KG model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d96fc04a",
   "metadata": {},
   "outputs": [],
   "source": [
    "device = 'cpu' # to convert to cpu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "58b83b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Top models\n",
    "# Consolidated meptapaths: /home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-18_17-24-55/train_m2vec_19f34_00000_0_batch_size=512,embedding_dim=128,lr=0.0029,num_negative_samples=5,walk_length=21,walks_per_node=12_2024-04-18_17-25-05/checkpoint_000000\n",
    "# entire elk: /home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-13_02-43-09/train_m2vec_178a4_00000_0_batch_size=256,embedding_dim=128,lr=0.0248,metapath=biolink_Gene_biolink_interacts_with_biolink_Gene_bio_2024-04-13_02-43-23/checkpoint_000000\n",
    "# \"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-18_17-24-55/train_m2vec_19f34_00000_0_batch_size=512,embedding_dim=128,lr=0.0029,num_negative_samples=5,walk_length=21,walks_per_node=12_2024-04-18_17-25-05/checkpoint_000000/checkpoint\"\n",
    "# precious_carp_5539: /home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-03_01-44-54/train_m2vec_4bf4d_00001_1_batch_size=128,embedding_dim=128,lr=0.0050,walk_length=12_2024-04-03_01-44-55/checkpoint_000000/checkpoint\n",
    "# (active) hilarious_airway_5834: /home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-14_17-21-49/train_m2vec_016cc_00000_0_batch_size=32,embedding_dim=128,lr=0.0104,metapath=biolink_Gene_biolink_interacts_with_biolink_Gene_biol_2024-04-14_17-22-01/checkpoint_000000\n",
    "\n",
    "chkpt_filepath = \"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-14_17-21-49/train_m2vec_016cc_00000_0_batch_size=32,embedding_dim=128,lr=0.0104,metapath=biolink_Gene_biolink_interacts_with_biolink_Gene_biol_2024-04-14_17-22-01/checkpoint_000000/checkpoint\"\n",
    "# chkpt_filepath = \"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/06_models/modspy-experiments/ray-reults/train_m2vec_2024-04-03_01-44-54/train_m2vec_4bf4d_00001_1_batch_size=128,embedding_dim=128,lr=0.0050,walk_length=12_2024-04-03_01-44-55/checkpoint_000000/checkpoint\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d758284c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing on cpu\n",
      "True\n",
      "1\n",
      "Tesla P100-PCIE-12GB\n",
      " \n",
      "        Metapath: [('biolink:Gene', 'biolink:interacts_with', 'biolink:Gene'), ('biolink:Gene', 'biolink:orthologous_to', 'biolink:Gene'), ('biolink:Gene', 'biolink:interacts_with', 'biolink:Gene')]\n",
      "        Total nodes: 862115\n",
      "        Total node types: 88\n",
      "\n",
      "        Total edges: 11412471\n",
      "        Total edge types: 289                \n",
      "        \n",
      "cpu\n",
      "MetaPath2VecLightningModule(\n",
      "  (model): MetaPath2Vec(559272, 128)\n",
      "  (val_precision): BinaryPrecision()\n",
      "  (val_recall): BinaryRecall()\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "model = MetaPath2VecLightningModule.load_from_checkpoint(chkpt_filepath).to(device)\n",
    "print(model.device)\n",
    "print(model.eval())\n",
    "graph = torch.load(model.hparams['network_filepath']).to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2331c0f7",
   "metadata": {},
   "source": [
    "<b>Creating dataset with node representation<b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "8af9b315",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5541, 256)\n",
      "(5541, 77)\n",
      "(5541, 333)\n",
      "Resultant array\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3412/3817260123.py:4: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  X = np.concatenate((model.model.embedding(torch.tensor(data_tensor[:,0])).detach().cpu().numpy(), model.model.embedding(torch.tensor(data_tensor[:,1])).detach().cpu().numpy()), axis=1)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "emb_0    0.185309\n",
       "emb_1    0.150688\n",
       "emb_2   -0.201403\n",
       "emb_3   -0.201173\n",
       "emb_4    0.227061\n",
       "Name: 0, dtype: object"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Querying individual 1st node's embeding\n",
      "[ 0.18530875  0.15068765 -0.20140305 -0.20117337  0.22706136]\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3412/3817260123.py:23: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
      "  print(model.model.embedding(torch.tensor(data_tensor[0,0])).detach().cpu().numpy()[:5])\n"
     ]
    }
   ],
   "source": [
    "if type(data_tensor) is torch.tensor:\n",
    "    data_tensor = data_tensor.detach().numpy()\n",
    "\n",
    "X = np.concatenate((model.model.embedding(torch.tensor(data_tensor[:,0])).detach().cpu().numpy(), model.model.embedding(torch.tensor(data_tensor[:,1])).detach().cpu().numpy()), axis=1)\n",
    "y = data_tensor[:,2]\n",
    "assert X.shape[0] == y.shape[0]\n",
    "print(X.shape)\n",
    "print(dataset_w_ninfo.shape)\n",
    "\n",
    "col_names = [f\"emb_{i}\" for i in range(0, X.shape[1])]\n",
    "emb_df = pd.DataFrame(X.astype(np.float64), columns=col_names, dtype=np.float128)\n",
    "\n",
    "# Droping index for alignment\n",
    "dataset_w_ninfo = dataset_w_ninfo.reset_index(drop=True)\n",
    "emb_df = emb_df.reset_index(drop=True)\n",
    "\n",
    "dataset_w_emb = pd.concat([dataset_w_ninfo, emb_df], axis=1)\n",
    "print(dataset_w_emb.shape)\n",
    "\n",
    "print(\"Resultant array\")\n",
    "display(dataset_w_emb.loc[0,['emb_0','emb_1','emb_2','emb_3','emb_4']])\n",
    "print(\"\\nQuerying individual 1st node's embeding\")\n",
    "print(model.model.embedding(torch.tensor(data_tensor[0,0])).detach().cpu().numpy()[:5])\n",
    "print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "815e7418",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset_w_emb.to_csv('./data/05_model_input/2024-06-12-merged-dataset-with_node_embedding.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f87a33",
   "metadata": {},
   "source": [
    "#### Creating ModifierDataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3099d863",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from tqdm import tqdm\n",
    "\n",
    "def verify_heterodata_construction(data: HeteroData, edges_ddf, node_ids):\n",
    "    edge_type_to_chk = 'biolink:interacts_with'\n",
    "    \n",
    "    # Access the edge index for the relation\n",
    "    edge_index = monarch['biolink:Gene', edge_type_to_chk, 'biolink:Gene'].edge_index    \n",
    "    src, dest = edge_index\n",
    "    \n",
    "    for node_idx in tqdm(node_ids):\n",
    "        graph_in_count = src[dest == node_idx].shape[0]\n",
    "        graph_out_count = dest[src == node_idx].shape[0]\n",
    "        \n",
    "        orig_in_count = edges_df[(edges_df['object_id']==node_idx) & (edges_df['predicate']==edge_type_to_chk)]['id'].compute().shape[0]\n",
    "        orig_out_count = edges_df[(edges_df['subject_id']==node_idx) & (edges_df['predicate']==edge_type_to_chk)]['id'].compute().shape[0]\n",
    "\n",
    "\n",
    "        try:\n",
    "            assert graph_in_count == orig_in_count\n",
    "            assert graph_out_count == orig_out_count\n",
    "        except AssertionError as e:\n",
    "            print(f\"AssertionError: {e}\")\n",
    "            print(f\"{graph_in_count} != {orig_in_count}\")\n",
    "            print('or')\n",
    "            print(f\"{graph_out_count} != {orig_out_count}\")\n",
    "\n",
    "        # # Print results\n",
    "        # print(\"Outgoing edges for node\", node_idx, \" (\", outgoing_edges.shape[0] ,\")\" \":\", outgoing_edges.tolist())\n",
    "        # print(\"Incoming edges for node\", node_idx, \" (\", incoming_edges.shape[0] ,\")\" \":\", incoming_edges.tolist())\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cec1e59d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "153\n"
     ]
    }
   ],
   "source": [
    "node_ids = list(set(data_t[:,0].tolist()))\n",
    "print(len(node_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "716bad9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/153 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 153/153 [12:20<00:00,  4.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Succesfully verified 153 genes in the network!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "verify_heterodata_construction(monarch, edges_df, node_ids)\n",
    "print(f\"Succesfully verified {len(node_ids)} genes in the network!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4441a0fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "\n",
    "class ModifierDataset(Dataset):\n",
    "    def __init__(self, filepath: str = None):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            data (Tensor): A tensor containing node pairs and their similarity label.\n",
    "                           Shape: [num_pairs, 3], where each row is (node1, node2, label).\n",
    "        \"\"\"\n",
    "        self.data = torch.load(filepath)\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        modifier, target, label = self.data[idx]\n",
    "        return modifier, target, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5966f8da",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = ModifierDataset(\"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/05_model_input/2024-03-31-merged-dataset.pt\")\n",
    "val_loader = DataLoader(dataset, batch_size=len(dataset), shuffle=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4dda1baf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(tensor(516790, dtype=torch.int32),\n",
       " tensor(518908, dtype=torch.int32),\n",
       " tensor(1, dtype=torch.int32))"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(iter(dataset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "027f5d1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([     0,      1,  81868, ..., 558429, 558431, 558747], dtype=int32)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = torch.load(\"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/05_model_input/2024-03-31-merged-dataset.pt\")\n",
    "np.unique(data.numpy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f0f5514",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.5 (ModSpy CUDA) Kernel",
   "language": "python",
   "name": "modspy-py311-cuda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
