{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3173e30c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rahit/projects/def-mtarailo/rahit/python_environments/py311/lib/python3.11/site-packages/dask/dataframe/_pyarrow_compat.py:17: FutureWarning: Minimal version of pyarrow will soon be increased to 14.0.1. You are using 12.0.1. Please consider upgrading.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import torch\n",
    "import optuna\n",
    "from comet_ml import Experiment\n",
    "from pytorch_lightning.loggers import CometLogger\n",
    "from torch_geometric.transforms import AddSelfLoops\n",
    "import wandb\n",
    "\n",
    "\n",
    "import dask\n",
    "import dask.dataframe as dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "607309ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#wandb.init(mode=\"offline\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f528f5c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "Total nodes: 862115\n",
      "Total node types: 88\n",
      "\n",
      "Total edges: 11412471\n",
      "Total edge types: 289                \n",
      "\n"
     ]
    }
   ],
   "source": [
    "monarch = torch.load('./data/05_model_input/2024-02-monarch_heterodata_v1.pt', map_location='cpu')\n",
    "\n",
    "print(\n",
    "    f\"\"\" \n",
    "Total nodes: {monarch.num_nodes}\n",
    "Total node types: {len(monarch.node_types)}\n",
    "\n",
    "Total edges: {monarch.num_edges}\n",
    "Total edge types: {len(monarch.edge_types)}                \n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "# # Adding self loops to avoid 1. nodes without any edge, 2. consider intragenic modifier\n",
    "# transform = AddSelfLoops()\n",
    "# monarch = transform(monarch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "78d727a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "jvl = pd.read_csv('./data/02_intermediate/jvl/annotated.csv', sep='\\t')\n",
    "olida = pd.read_csv('./data/02_intermediate/olida/annotated.tsv', sep='\\t')\n",
    "mtg = pd.read_csv('./data/02_intermediate/mtg/annotated.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b1bd8d82",
   "metadata": {},
   "outputs": [],
   "source": [
    "olida = olida[olida['Oligogenic Effect']!='Monogenic+Modifier']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9fb9b8bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Oligogenic Effect\n",
      "True Digenic                47\n",
      "Dual Molecular Diagnosis    28\n",
      "Name: count, dtype: int64\n",
      "(75, 47)\n"
     ]
    }
   ],
   "source": [
    "print(olida['Oligogenic Effect'].value_counts())\n",
    "print(olida.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f7feb16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding source column\n",
    "jvl['source'] = 'JVL'\n",
    "olida['source'] = 'OLIDA'\n",
    "mtg['source'] = 'MTG'\n",
    "\n",
    "# Add species information\n",
    "jvl['species'] = 'Homo sapiens'\n",
    "olida['species'] = 'Homo sapiens'\n",
    "mtg['species'] = 'Caenorhabditis elegans'\n",
    "\n",
    "# Classification labels\n",
    "jvl['is_modifier'] = 1\n",
    "olida['is_modifier'] = 0\n",
    "\n",
    "# Rename columns to indicate same information\n",
    "jvl.rename(columns={'QueryGene': 'target_gene', 'SuppressorGene': 'modifier_gene'}, inplace=True)\n",
    "olida.rename(columns={'gene_a': 'target_gene', 'gene_b': 'modifier_gene'}, inplace=True)\n",
    "mtg.rename(columns={'gene_symbol': 'modifier_gene', 'target_gene_symbol': 'target_gene'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "44ea6e52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8337, 219)\n",
      "source\n",
      "MTG      7330\n",
      "JVL       932\n",
      "OLIDA      75\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "dataset_df = pd.concat([jvl, olida, mtg])\n",
    "print(dataset_df.shape)\n",
    "print(dataset_df['source'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3548a28f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['PubmedID',\n",
       " 'Category',\n",
       " 'Tissue',\n",
       " 'target_gene',\n",
       " 'QueryFunction',\n",
       " 'QueryMutation',\n",
       " 'QueryType',\n",
       " 'modifier_gene',\n",
       " 'SuppressorFunction',\n",
       " 'SuppressorMutation',\n",
       " 'SNP_ID',\n",
       " 'SuppressorType',\n",
       " 'EffectSize',\n",
       " 'Disease',\n",
       " 'DiseaseSubType',\n",
       " 'CellLineIdentified',\n",
       " 'ModelSystemValidated',\n",
       " 'Drugs',\n",
       " 'target_GOs',\n",
       " 'target_GOs_count',\n",
       " 'modifier_GOs',\n",
       " 'modifier_GOs_count',\n",
       " 'target_POs',\n",
       " 'target_POs_count',\n",
       " 'modifier_POs',\n",
       " 'modifier_POs_count',\n",
       " 'target_DOs',\n",
       " 'target_DOs_count',\n",
       " 'modifier_DOs',\n",
       " 'modifier_DOs_count',\n",
       " 'source',\n",
       " 'species',\n",
       " 'is_modifier',\n",
       " 'Entry Id',\n",
       " 'Genes',\n",
       " 'Genes Relationship',\n",
       " 'Protein Interactions',\n",
       " 'Common Pathways',\n",
       " 'GENEmeta_x',\n",
       " 'Oligogenic variant combinations',\n",
       " 'olida_id',\n",
       " 'OLIDA ID',\n",
       " 'Omim Id',\n",
       " 'Diseases',\n",
       " 'Oligogenic Effect',\n",
       " 'Ethnicity',\n",
       " 'References',\n",
       " 'Associated Variants',\n",
       " 'FAMmanual',\n",
       " 'STATmanual',\n",
       " 'STATknowledge',\n",
       " 'STATmeta',\n",
       " 'GENEmanual',\n",
       " 'GENEmanual harmonized',\n",
       " 'GENEknowledge',\n",
       " 'GENEmeta_y',\n",
       " 'VARmanual',\n",
       " 'VARknowledge',\n",
       " 'VARmeta',\n",
       " 'FUNmanual',\n",
       " 'FUNmeta',\n",
       " 'FINALmanual',\n",
       " 'FINALmeta',\n",
       " 'gene_list',\n",
       " 'gene_list_len',\n",
       " 'gene_pairs',\n",
       " 'Sample',\n",
       " 'chromosome',\n",
       " 'position',\n",
       " 'ref',\n",
       " 'alt',\n",
       " 'wormbase_gene_id',\n",
       " 'gene_start_position',\n",
       " 'gene_end_position',\n",
       " 'strain',\n",
       " 'variant_id',\n",
       " 'target_gene_id',\n",
       " 'go_Resnik_max_similarity',\n",
       " 'go_Resnik_BMA_similarity',\n",
       " 'go_Resnik_avg_similarity',\n",
       " 'go_Lin_max_similarity',\n",
       " 'go_Lin_BMA_similarity',\n",
       " 'go_Lin_avg_similarity',\n",
       " 'go_Jiang-Conrath_max_similarity',\n",
       " 'go_Jiang-Conrath_BMA_similarity',\n",
       " 'go_Jiang-Conrath_avg_similarity',\n",
       " 'go_SimRel_max_similarity',\n",
       " 'go_SimRel_BMA_similarity',\n",
       " 'go_SimRel_avg_similarity',\n",
       " 'go_SimIC_max_similarity',\n",
       " 'go_SimIC_BMA_similarity',\n",
       " 'go_SimIC_avg_similarity',\n",
       " 'go_G-SESAME_max_similarity',\n",
       " 'go_G-SESAME_BMA_similarity',\n",
       " 'go_G-SESAME_avg_similarity',\n",
       " 'go_SimICND_max_similarity',\n",
       " 'go_SimICND_BMA_similarity',\n",
       " 'go_SimICND_avg_similarity',\n",
       " 'go_SimICNP_max_similarity',\n",
       " 'go_SimICNP_BMA_similarity',\n",
       " 'go_SimICNP_avg_similarity',\n",
       " 'wpo_Resnik_max_similarity',\n",
       " 'wpo_Resnik_BMA_similarity',\n",
       " 'wpo_Resnik_avg_similarity',\n",
       " 'wpo_Lin_max_similarity',\n",
       " 'wpo_Lin_BMA_similarity',\n",
       " 'wpo_Lin_avg_similarity',\n",
       " 'wpo_Jiang-Conrath_max_similarity',\n",
       " 'wpo_Jiang-Conrath_BMA_similarity',\n",
       " 'wpo_Jiang-Conrath_avg_similarity',\n",
       " 'wpo_SimRel_max_similarity',\n",
       " 'wpo_SimRel_BMA_similarity',\n",
       " 'wpo_SimRel_avg_similarity',\n",
       " 'wpo_SimIC_max_similarity',\n",
       " 'wpo_SimIC_BMA_similarity',\n",
       " 'wpo_SimIC_avg_similarity',\n",
       " 'wpo_G-SESAME_max_similarity',\n",
       " 'wpo_G-SESAME_BMA_similarity',\n",
       " 'wpo_G-SESAME_avg_similarity',\n",
       " 'wpo_SimICND_max_similarity',\n",
       " 'wpo_SimICND_BMA_similarity',\n",
       " 'wpo_SimICND_avg_similarity',\n",
       " 'wpo_SimICNP_max_similarity',\n",
       " 'wpo_SimICNP_BMA_similarity',\n",
       " 'wpo_SimICNP_avg_similarity',\n",
       " 'wpo_not_Resnik_max_similarity',\n",
       " 'wpo_not_Resnik_BMA_similarity',\n",
       " 'wpo_not_Resnik_avg_similarity',\n",
       " 'wpo_not_Lin_max_similarity',\n",
       " 'wpo_not_Lin_BMA_similarity',\n",
       " 'wpo_not_Lin_avg_similarity',\n",
       " 'wpo_not_Jiang-Conrath_max_similarity',\n",
       " 'wpo_not_Jiang-Conrath_BMA_similarity',\n",
       " 'wpo_not_Jiang-Conrath_avg_similarity',\n",
       " 'wpo_not_SimRel_max_similarity',\n",
       " 'wpo_not_SimRel_BMA_similarity',\n",
       " 'wpo_not_SimRel_avg_similarity',\n",
       " 'wpo_not_SimIC_max_similarity',\n",
       " 'wpo_not_SimIC_BMA_similarity',\n",
       " 'wpo_not_SimIC_avg_similarity',\n",
       " 'wpo_not_G-SESAME_max_similarity',\n",
       " 'wpo_not_G-SESAME_BMA_similarity',\n",
       " 'wpo_not_G-SESAME_avg_similarity',\n",
       " 'wpo_not_SimICND_max_similarity',\n",
       " 'wpo_not_SimICND_BMA_similarity',\n",
       " 'wpo_not_SimICND_avg_similarity',\n",
       " 'wpo_not_SimICNP_max_similarity',\n",
       " 'wpo_not_SimICNP_BMA_similarity',\n",
       " 'wpo_not_SimICNP_avg_similarity',\n",
       " 'protein1',\n",
       " 'protein2',\n",
       " 'neighborhood',\n",
       " 'neighborhood_transferred',\n",
       " 'fusion',\n",
       " 'cooccurence',\n",
       " 'homology',\n",
       " 'coexpression',\n",
       " 'coexpression_transferred',\n",
       " 'experiments',\n",
       " 'experiments_transferred',\n",
       " 'database',\n",
       " 'database_transferred',\n",
       " 'textmining',\n",
       " 'textmining_transferred',\n",
       " 'combined_score',\n",
       " 'go_n_common_ancestors_max',\n",
       " 'go_n_common_ancestors_avg',\n",
       " 'go_n_common_ancestors_bma',\n",
       " 'go_n_union_ancestors_max',\n",
       " 'go_n_union_ancestors_avg',\n",
       " 'go_n_union_ancestors_bma',\n",
       " 'go_batet_max',\n",
       " 'go_batet_avg',\n",
       " 'go_batet_bma',\n",
       " 'go_batet_log_max',\n",
       " 'go_batet_log_avg',\n",
       " 'go_batet_log_bma',\n",
       " 'go_resnik_max',\n",
       " 'go_resnik_avg',\n",
       " 'go_resnik_bma',\n",
       " 'go_resnik_scaled_max',\n",
       " 'go_resnik_scaled_avg',\n",
       " 'go_resnik_scaled_bma',\n",
       " 'go_lin_max',\n",
       " 'go_lin_avg',\n",
       " 'go_lin_bma',\n",
       " 'go_jiang_max',\n",
       " 'go_jiang_avg',\n",
       " 'go_jiang_bma',\n",
       " 'go_jiang_seco_max',\n",
       " 'go_jiang_seco_avg',\n",
       " 'go_jiang_seco_bma',\n",
       " 'wpo_n_common_ancestors_max',\n",
       " 'wpo_n_common_ancestors_avg',\n",
       " 'wpo_n_common_ancestors_bma',\n",
       " 'wpo_n_union_ancestors_max',\n",
       " 'wpo_n_union_ancestors_avg',\n",
       " 'wpo_n_union_ancestors_bma',\n",
       " 'wpo_batet_max',\n",
       " 'wpo_batet_avg',\n",
       " 'wpo_batet_bma',\n",
       " 'wpo_batet_log_max',\n",
       " 'wpo_batet_log_avg',\n",
       " 'wpo_batet_log_bma',\n",
       " 'wpo_resnik_max',\n",
       " 'wpo_resnik_avg',\n",
       " 'wpo_resnik_bma',\n",
       " 'wpo_resnik_scaled_max',\n",
       " 'wpo_resnik_scaled_avg',\n",
       " 'wpo_resnik_scaled_bma',\n",
       " 'wpo_lin_max',\n",
       " 'wpo_lin_avg',\n",
       " 'wpo_lin_bma',\n",
       " 'wpo_jiang_max',\n",
       " 'wpo_jiang_avg',\n",
       " 'wpo_jiang_bma',\n",
       " 'wpo_jiang_seco_max',\n",
       " 'wpo_jiang_seco_avg',\n",
       " 'wpo_jiang_seco_bma']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(dataset_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7d34fcd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_GOs</th>\n",
       "      <th>modifier_GOs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0005080,GO:0005200,GO:0005516,GO:0007010,GO...</td>\n",
       "      <td>GO:0000977,GO:0003682,GO:0003712,GO:0003712,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0071407,GO:0004497,GO:0004497,GO:0004497,GO...</td>\n",
       "      <td>GO:0051897,GO:0070374,GO:0001725,GO:0005884,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0005243,GO:0005243,GO:0005509,GO:0005515,GO...</td>\n",
       "      <td>GO:0006883,GO:0004252,GO:0004252,GO:0017080,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0005515,GO:0005515,GO:0005515,GO:0005515,GO...</td>\n",
       "      <td>GO:0002153,GO:0002153,GO:0006357,GO:1990904,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0003779,GO:0005515,GO:0005515,GO:0005515,GO...</td>\n",
       "      <td>GO:0003723,GO:0005164,GO:0005515,GO:0005515,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>GO:0005524,GO:0016887,GO:0120020,GO:0005515,GO...</td>\n",
       "      <td>GO:0003723,GO:0004252,GO:0005515,GO:0005515,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>GO:0001540,GO:0001540,GO:0001540,GO:0005041,GO...</td>\n",
       "      <td>GO:0003723,GO:0004252,GO:0005515,GO:0005515,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>GO:0000122,GO:0010628,GO:0010628,GO:0010628,GO...</td>\n",
       "      <td>GO:0000976,GO:0003682,GO:0005515,GO:0005515,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>GO:0030509,GO:0005179,GO:0005515,GO:0005515,GO...</td>\n",
       "      <td>GO:0003677,GO:0003714,GO:0005515,GO:0005515,GO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>GO:0005802,GO:0005802,GO:0016323,GO:0031410,GO...</td>\n",
       "      <td>GO:0016328,GO:0002162,GO:0002162,GO:0003779,GO...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            target_GOs  \\\n",
       "1    GO:0005080,GO:0005200,GO:0005516,GO:0007010,GO...   \n",
       "3    GO:0071407,GO:0004497,GO:0004497,GO:0004497,GO...   \n",
       "4    GO:0005243,GO:0005243,GO:0005509,GO:0005515,GO...   \n",
       "6    GO:0005515,GO:0005515,GO:0005515,GO:0005515,GO...   \n",
       "10   GO:0003779,GO:0005515,GO:0005515,GO:0005515,GO...   \n",
       "..                                                 ...   \n",
       "129  GO:0005524,GO:0016887,GO:0120020,GO:0005515,GO...   \n",
       "130  GO:0001540,GO:0001540,GO:0001540,GO:0005041,GO...   \n",
       "135  GO:0000122,GO:0010628,GO:0010628,GO:0010628,GO...   \n",
       "136  GO:0030509,GO:0005179,GO:0005515,GO:0005515,GO...   \n",
       "139  GO:0005802,GO:0005802,GO:0016323,GO:0031410,GO...   \n",
       "\n",
       "                                          modifier_GOs  \n",
       "1    GO:0000977,GO:0003682,GO:0003712,GO:0003712,GO...  \n",
       "3    GO:0051897,GO:0070374,GO:0001725,GO:0005884,GO...  \n",
       "4    GO:0006883,GO:0004252,GO:0004252,GO:0017080,GO...  \n",
       "6    GO:0002153,GO:0002153,GO:0006357,GO:1990904,GO...  \n",
       "10   GO:0003723,GO:0005164,GO:0005515,GO:0005515,GO...  \n",
       "..                                                 ...  \n",
       "129  GO:0003723,GO:0004252,GO:0005515,GO:0005515,GO...  \n",
       "130  GO:0003723,GO:0004252,GO:0005515,GO:0005515,GO...  \n",
       "135  GO:0000976,GO:0003682,GO:0005515,GO:0005515,GO...  \n",
       "136  GO:0003677,GO:0003714,GO:0005515,GO:0005515,GO...  \n",
       "139  GO:0016328,GO:0002162,GO:0002162,GO:0003779,GO...  \n",
       "\n",
       "[75 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "olida[['target_GOs','modifier_GOs']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "abfe5a08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>species</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>modifier_gene</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>PubmedID</th>\n",
       "      <th>Category</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>QueryFunction</th>\n",
       "      <th>QueryMutation</th>\n",
       "      <th>...</th>\n",
       "      <th>wpo_resnik_scaled_bma</th>\n",
       "      <th>wpo_lin_max</th>\n",
       "      <th>wpo_lin_avg</th>\n",
       "      <th>wpo_lin_bma</th>\n",
       "      <th>wpo_jiang_max</th>\n",
       "      <th>wpo_jiang_avg</th>\n",
       "      <th>wpo_jiang_bma</th>\n",
       "      <th>wpo_jiang_seco_max</th>\n",
       "      <th>wpo_jiang_seco_avg</th>\n",
       "      <th>wpo_jiang_seco_bma</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>CASP7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>27358062.0</td>\n",
       "      <td>Patients</td>\n",
       "      <td>Neuron</td>\n",
       "      <td>Lipid and sterol biosynthesis &amp; transport</td>\n",
       "      <td>C112R/C112R</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>HBB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24116184.0</td>\n",
       "      <td>Patients</td>\n",
       "      <td>-</td>\n",
       "      <td>Lipid and sterol biosynthesis &amp; transport</td>\n",
       "      <td>C112R/?</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>KL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>30867273.0</td>\n",
       "      <td>Patients</td>\n",
       "      <td>Neuron</td>\n",
       "      <td>Lipid and sterol biosynthesis &amp; transport</td>\n",
       "      <td>C112R/?</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>KL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>32282020.0</td>\n",
       "      <td>Patients</td>\n",
       "      <td>Neuron</td>\n",
       "      <td>Lipid and sterol biosynthesis &amp; transport</td>\n",
       "      <td>C112R/?</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATR</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>23284306.0</td>\n",
       "      <td>Cells</td>\n",
       "      <td>-</td>\n",
       "      <td>DNA replication and repair;Signaling &amp; stress ...</td>\n",
       "      <td>silencing/silencing</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 219 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  source       species target_gene modifier_gene  is_modifier    PubmedID  \\\n",
       "0    JVL  Homo sapiens        APOE         CASP7          1.0  27358062.0   \n",
       "1    JVL  Homo sapiens        APOE           HBB          1.0  24116184.0   \n",
       "2    JVL  Homo sapiens        APOE            KL          1.0  30867273.0   \n",
       "3    JVL  Homo sapiens        APOE            KL          1.0  32282020.0   \n",
       "4    JVL  Homo sapiens         ATR          ETV1          1.0  23284306.0   \n",
       "\n",
       "   Category  Tissue                                      QueryFunction  \\\n",
       "0  Patients  Neuron          Lipid and sterol biosynthesis & transport   \n",
       "1  Patients       -          Lipid and sterol biosynthesis & transport   \n",
       "2  Patients  Neuron          Lipid and sterol biosynthesis & transport   \n",
       "3  Patients  Neuron          Lipid and sterol biosynthesis & transport   \n",
       "4     Cells       -  DNA replication and repair;Signaling & stress ...   \n",
       "\n",
       "         QueryMutation  ... wpo_resnik_scaled_bma wpo_lin_max wpo_lin_avg  \\\n",
       "0          C112R/C112R  ...                   NaN         NaN         NaN   \n",
       "1              C112R/?  ...                   NaN         NaN         NaN   \n",
       "2              C112R/?  ...                   NaN         NaN         NaN   \n",
       "3              C112R/?  ...                   NaN         NaN         NaN   \n",
       "4  silencing/silencing  ...                   NaN         NaN         NaN   \n",
       "\n",
       "  wpo_lin_bma wpo_jiang_max wpo_jiang_avg wpo_jiang_bma wpo_jiang_seco_max  \\\n",
       "0         NaN           NaN           NaN           NaN                NaN   \n",
       "1         NaN           NaN           NaN           NaN                NaN   \n",
       "2         NaN           NaN           NaN           NaN                NaN   \n",
       "3         NaN           NaN           NaN           NaN                NaN   \n",
       "4         NaN           NaN           NaN           NaN                NaN   \n",
       "\n",
       "  wpo_jiang_seco_avg wpo_jiang_seco_bma  \n",
       "0                NaN                NaN  \n",
       "1                NaN                NaN  \n",
       "2                NaN                NaN  \n",
       "3                NaN                NaN  \n",
       "4                NaN                NaN  \n",
       "\n",
       "[5 rows x 219 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reordering important columns\n",
    "desired_first_columns = ['source', 'species', 'target_gene', 'modifier_gene', 'is_modifier'] \n",
    "remaining_columns = [col for col in dataset_df.columns if col not in desired_first_columns]\n",
    "new_column_order = desired_first_columns + remaining_columns\n",
    "dataset_df = dataset_df[new_column_order]\n",
    "dataset_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ee888365",
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_df = dd.read_parquet('./data/02_intermediate/monarch/nodes_with_type_idx')  \n",
    "edges_df = dd.read_parquet('./data/02_intermediate/monarch/edges_pre_df_reduction_v2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fce840d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "862115\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>name</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>in_taxon_label</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>APO:0000017</th>\n",
       "      <td>biolink:PhenotypicFeature</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000001</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>entity</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000002</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>continuant</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000003</th>\n",
       "      <td>biolink:BiologicalProcessOrActivity</td>\n",
       "      <td>occurrent</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFO:0000004</th>\n",
       "      <td>biolink:NamedThing</td>\n",
       "      <td>independent continuant</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        category                    name  \\\n",
       "id                                                                         \n",
       "APO:0000017            biolink:PhenotypicFeature                    <NA>   \n",
       "BFO:0000001                   biolink:NamedThing                  entity   \n",
       "BFO:0000002                   biolink:NamedThing              continuant   \n",
       "BFO:0000003  biolink:BiologicalProcessOrActivity               occurrent   \n",
       "BFO:0000004                   biolink:NamedThing  independent continuant   \n",
       "\n",
       "            in_taxon in_taxon_label symbol  type_index  \n",
       "id                                                      \n",
       "APO:0000017     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000001     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000002     <NA>           <NA>   <NA>           1  \n",
       "BFO:0000003     <NA>           <NA>   <NA>           0  \n",
       "BFO:0000004     <NA>           <NA>   <NA>           2  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(nodes_df.index.shape[0].compute())\n",
    "display(nodes_df.head())\n",
    "\n",
    "# Use dataset's source and target gene id to pull the `type_index` column From `nodes_df``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6138482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>name</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>in_taxon_label</th>\n",
       "      <th>symbol</th>\n",
       "      <th>type_index</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HGNC:613</th>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>APOE</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>516790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              category  name        in_taxon in_taxon_label symbol  type_index\n",
       "id                                                                            \n",
       "HGNC:613  biolink:Gene  APOE  NCBITaxon:9606   Homo sapiens   APOE      516790"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_df[(nodes_df['category']=='biolink:Gene') & (nodes_df['symbol']=='APOE') & (nodes_df['in_taxon_label']=='Homo sapiens')].compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1cf1356e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                              <NA>\n",
       "1                      Homo sapiens\n",
       "2                      Mus musculus\n",
       "3                     Gallus gallus\n",
       "4                        Bos taurus\n",
       "5                 Rattus norvegicus\n",
       "0           Drosophila melanogaster\n",
       "1                        Sus scrofa\n",
       "2            Canis lupus familiaris\n",
       "3          Dictyostelium discoideum\n",
       "4         Schizosaccharomyces pombe\n",
       "5    Saccharomyces cerevisiae S288C\n",
       "6            Caenorhabditis elegans\n",
       "7                    Xenopus laevis\n",
       "8                Xenopus tropicalis\n",
       "9                       Danio rerio\n",
       "Name: in_taxon_label, dtype: string"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_df['in_taxon_label'].unique().compute()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c57d3f",
   "metadata": {},
   "source": [
    "### Merging Node Index from Monarch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b807404f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>species</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>modifier_gene</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>CASP7</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516790</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>518908.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>HBB</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516790</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>526845.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>KL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516790</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>529553.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>APOE</td>\n",
       "      <td>KL</td>\n",
       "      <td>1.0</td>\n",
       "      <td>516790</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>529553.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JVL</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ATR</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>517533</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>523802.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8332</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>C39D10.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>94433.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8333</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>catp-5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>96963.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8334</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>tmc-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>93262.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8335</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>tbb-4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>86304.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8336</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>rrc-1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>biolink:Gene</td>\n",
       "      <td>NCBITaxon:6239</td>\n",
       "      <td>89182.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8337 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     source                 species target_gene modifier_gene  is_modifier  \\\n",
       "0       JVL            Homo sapiens        APOE         CASP7          1.0   \n",
       "1       JVL            Homo sapiens        APOE           HBB          1.0   \n",
       "2       JVL            Homo sapiens        APOE            KL          1.0   \n",
       "3       JVL            Homo sapiens        APOE            KL          1.0   \n",
       "4       JVL            Homo sapiens         ATR          ETV1          1.0   \n",
       "...     ...                     ...         ...           ...          ...   \n",
       "8332    MTG  Caenorhabditis elegans       zyg-1      C39D10.7          0.0   \n",
       "8333    MTG  Caenorhabditis elegans       zyg-1        catp-5          0.0   \n",
       "8334    MTG  Caenorhabditis elegans       zyg-1         tmc-2          0.0   \n",
       "8335    MTG  Caenorhabditis elegans       zyg-1         tbb-4          0.0   \n",
       "8336    MTG  Caenorhabditis elegans       zyg-1         rrc-1          0.0   \n",
       "\n",
       "      target_type_index      category        in_taxon  modifier_type_index  \n",
       "0                516790  biolink:Gene  NCBITaxon:9606             518908.0  \n",
       "1                516790  biolink:Gene  NCBITaxon:9606             526845.0  \n",
       "2                516790  biolink:Gene  NCBITaxon:9606             529553.0  \n",
       "3                516790  biolink:Gene  NCBITaxon:9606             529553.0  \n",
       "4                517533  biolink:Gene  NCBITaxon:9606             523802.0  \n",
       "...                 ...           ...             ...                  ...  \n",
       "8332              86692  biolink:Gene  NCBITaxon:6239              94433.0  \n",
       "8333              86692  biolink:Gene  NCBITaxon:6239              96963.0  \n",
       "8334              86692  biolink:Gene  NCBITaxon:6239              93262.0  \n",
       "8335              86692  biolink:Gene  NCBITaxon:6239              86304.0  \n",
       "8336              86692  biolink:Gene  NCBITaxon:6239              89182.0  \n",
       "\n",
       "[8337 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nodes_pdf = nodes_df.compute()\n",
    "_dataset = dataset_df.iloc[:,:5].merge(nodes_pdf, how='left', left_on=['species','target_gene'], right_on=['in_taxon_label','symbol'])\n",
    "_dataset.rename(columns={'type_index': 'target_type_index'}, inplace=True)\n",
    "_dataset.drop(columns=['category','name','in_taxon_label','in_taxon','symbol'], inplace=True)\n",
    "dataset_w_ninfo = _dataset.merge(nodes_pdf, how='left', left_on=['species','modifier_gene'], right_on=['in_taxon_label','symbol'])\n",
    "dataset_w_ninfo.rename(columns={'type_index': 'modifier_type_index'}, inplace=True)\n",
    "dataset_w_ninfo.drop(columns=['name','in_taxon_label','symbol'], inplace=True)\n",
    "del(_dataset)\n",
    "del(nodes_pdf)\n",
    "dataset_w_ninfo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "324d864c",
   "metadata": {},
   "source": [
    "### Did we detect all IDs for the genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0d3ff36f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "print(dataset_w_ninfo['target_type_index'].unique().shape == dataset_w_ninfo['target_gene'].unique().shape)\n",
    "print(dataset_w_ninfo['modifier_type_index'].unique().shape == dataset_w_ninfo['modifier_gene'].unique().shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ce6be17a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "253 genes are missing/not recognized from Monarch\n"
     ]
    }
   ],
   "source": [
    "print(f\"{dataset_w_ninfo['modifier_gene'].unique().shape[0] - dataset_w_ninfo['modifier_type_index'].unique().shape[0]} genes are missing/not recognized from Monarch\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e5093a1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unrecognized Modifiers:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>species</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>modifier_gene</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1068</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2976</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>apc-17</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4513</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6979</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     source                 species target_gene modifier_gene  is_modifier  \\\n",
       "1068    MTG  Caenorhabditis elegans       zyg-1         mat-2          1.0   \n",
       "2976    MTG  Caenorhabditis elegans       zyg-1        apc-17          1.0   \n",
       "4513    MTG  Caenorhabditis elegans       zyg-1         mat-2          1.0   \n",
       "6979    MTG  Caenorhabditis elegans       zyg-1         mat-2          1.0   \n",
       "\n",
       "      target_type_index category in_taxon  modifier_type_index  \n",
       "1068              86692     <NA>     <NA>                  NaN  \n",
       "2976              86692     <NA>     <NA>                  NaN  \n",
       "4513              86692     <NA>     <NA>                  NaN  \n",
       "6979              86692     <NA>     <NA>                  NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Unrecognized Modifiers:\")\n",
    "dataset_w_ninfo[(dataset_w_ninfo['modifier_type_index'].isna() & (dataset_w_ninfo['is_modifier']==1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8a6cc351",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing indecies\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>species</th>\n",
       "      <th>target_gene</th>\n",
       "      <th>modifier_gene</th>\n",
       "      <th>is_modifier</th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>category</th>\n",
       "      <th>in_taxon</th>\n",
       "      <th>modifier_type_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1018</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>T20B6.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1037</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>Y51A2D.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1068</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>mat-2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1086</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>Y71H2AM.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1117</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>C45E5.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8255</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>F36H2.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8269</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>BE0003N10.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8321</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>F59D12.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8324</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>R11G1.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8331</th>\n",
       "      <td>MTG</td>\n",
       "      <td>Caenorhabditis elegans</td>\n",
       "      <td>zyg-1</td>\n",
       "      <td>F48E3.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>86692</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>379 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     source                 species target_gene modifier_gene  is_modifier  \\\n",
       "1018    MTG  Caenorhabditis elegans       zyg-1       T20B6.3          0.0   \n",
       "1037    MTG  Caenorhabditis elegans       zyg-1      Y51A2D.7          0.0   \n",
       "1068    MTG  Caenorhabditis elegans       zyg-1         mat-2          1.0   \n",
       "1086    MTG  Caenorhabditis elegans       zyg-1    Y71H2AM.13          0.0   \n",
       "1117    MTG  Caenorhabditis elegans       zyg-1       C45E5.4          0.0   \n",
       "...     ...                     ...         ...           ...          ...   \n",
       "8255    MTG  Caenorhabditis elegans       zyg-1       F36H2.3          0.0   \n",
       "8269    MTG  Caenorhabditis elegans       zyg-1   BE0003N10.3          0.0   \n",
       "8321    MTG  Caenorhabditis elegans       zyg-1      F59D12.1          0.0   \n",
       "8324    MTG  Caenorhabditis elegans       zyg-1       R11G1.6          0.0   \n",
       "8331    MTG  Caenorhabditis elegans       zyg-1       F48E3.8          0.0   \n",
       "\n",
       "      target_type_index category in_taxon  modifier_type_index  \n",
       "1018              86692     <NA>     <NA>                  NaN  \n",
       "1037              86692     <NA>     <NA>                  NaN  \n",
       "1068              86692     <NA>     <NA>                  NaN  \n",
       "1086              86692     <NA>     <NA>                  NaN  \n",
       "1117              86692     <NA>     <NA>                  NaN  \n",
       "...                 ...      ...      ...                  ...  \n",
       "8255              86692     <NA>     <NA>                  NaN  \n",
       "8269              86692     <NA>     <NA>                  NaN  \n",
       "8321              86692     <NA>     <NA>                  NaN  \n",
       "8324              86692     <NA>     <NA>                  NaN  \n",
       "8331              86692     <NA>     <NA>                  NaN  \n",
       "\n",
       "[379 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f\"Missing indecies\")\n",
    "dataset_w_ninfo[(dataset_w_ninfo['modifier_type_index'].isna()) | (dataset_w_ninfo['target_type_index'].isna())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b02ae79a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Do we now have proper indecies?\n",
      "Target genes: True\n",
      "Modifier genes: True\n"
     ]
    }
   ],
   "source": [
    "# Dropping these NA rows\n",
    "dataset_w_ninfo.dropna(subset=['in_taxon','modifier_type_index'], inplace=True)\n",
    "\n",
    "print(f\"Do we now have proper indecies?\")\n",
    "print(f\"Target genes: {dataset_w_ninfo['target_type_index'].unique().shape == dataset_w_ninfo['target_gene'].unique().shape}\")\n",
    "print(f\"Modifier genes: {dataset_w_ninfo['modifier_type_index'].unique().shape == dataset_w_ninfo['modifier_gene'].unique().shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41567b04",
   "metadata": {},
   "source": [
    "### Adding target-modifier pairs as edges of Monarch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ed40cdf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_type_index</th>\n",
       "      <th>modifier_type_index</th>\n",
       "      <th>is_modifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>516790</td>\n",
       "      <td>518908</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>516790</td>\n",
       "      <td>526845</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>516790</td>\n",
       "      <td>529553</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>516790</td>\n",
       "      <td>529553</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>517533</td>\n",
       "      <td>523802</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8332</th>\n",
       "      <td>86692</td>\n",
       "      <td>94433</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8333</th>\n",
       "      <td>86692</td>\n",
       "      <td>96963</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8334</th>\n",
       "      <td>86692</td>\n",
       "      <td>93262</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8335</th>\n",
       "      <td>86692</td>\n",
       "      <td>86304</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8336</th>\n",
       "      <td>86692</td>\n",
       "      <td>89182</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7958 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      target_type_index  modifier_type_index  is_modifier\n",
       "0                516790               518908          1.0\n",
       "1                516790               526845          1.0\n",
       "2                516790               529553          1.0\n",
       "3                516790               529553          1.0\n",
       "4                517533               523802          1.0\n",
       "...                 ...                  ...          ...\n",
       "8332              86692                94433          0.0\n",
       "8333              86692                96963          0.0\n",
       "8334              86692                93262          0.0\n",
       "8335              86692                86304          0.0\n",
       "8336              86692                89182          0.0\n",
       "\n",
       "[7958 rows x 3 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_w_ninfo['modifier_type_index'] = dataset_w_ninfo['modifier_type_index'].astype(int)\n",
    "dataset_w_ninfo[['target_type_index','modifier_type_index','is_modifier']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d92496ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[516790, 518908,      1],\n",
       "        [516790, 526845,      1],\n",
       "        [516790, 529553,      1],\n",
       "        ...,\n",
       "        [ 86692,  93262,      0],\n",
       "        [ 86692,  86304,      0],\n",
       "        [ 86692,  89182,      0]], dtype=torch.int32)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert DataFrame to tensor\n",
    "dataset_arr = dataset_w_ninfo[['target_type_index','modifier_type_index','is_modifier']].to_numpy()\n",
    "data_t = torch.from_numpy(dataset_arr).to(torch.int)\n",
    "data_t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "25070552",
   "metadata": {},
   "outputs": [],
   "source": [
    "# torch.save(data_t, './data/05_model_input/2024-03-31-merged-dataset.pt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3099d863",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from tqdm import tqdm\n",
    "\n",
    "def verify_heterodata_construction(data: HeteroData, edges_ddf, node_ids):\n",
    "    edge_type_to_chk = 'biolink:interacts_with'\n",
    "    \n",
    "    # Access the edge index for the relation\n",
    "    edge_index = monarch['biolink:Gene', edge_type_to_chk, 'biolink:Gene'].edge_index    \n",
    "    src, dest = edge_index\n",
    "    \n",
    "    for node_idx in tqdm(node_ids):\n",
    "        graph_in_count = src[dest == node_idx].shape[0]\n",
    "        graph_out_count = dest[src == node_idx].shape[0]\n",
    "        \n",
    "        orig_in_count = edges_df[(edges_df['object_id']==node_idx) & (edges_df['predicate']==edge_type_to_chk)]['id'].compute().shape[0]\n",
    "        orig_out_count = edges_df[(edges_df['subject_id']==node_idx) & (edges_df['predicate']==edge_type_to_chk)]['id'].compute().shape[0]\n",
    "\n",
    "\n",
    "        try:\n",
    "            assert graph_in_count == orig_in_count\n",
    "            assert graph_out_count == orig_out_count\n",
    "        except AssertionError as e:\n",
    "            print(f\"AssertionError: {e}\")\n",
    "            print(f\"{graph_in_count} != {orig_in_count}\")\n",
    "            print('or')\n",
    "            print(f\"{graph_out_count} != {orig_out_count}\")\n",
    "\n",
    "        # # Print results\n",
    "        # print(\"Outgoing edges for node\", node_idx, \" (\", outgoing_edges.shape[0] ,\")\" \":\", outgoing_edges.tolist())\n",
    "        # print(\"Incoming edges for node\", node_idx, \" (\", incoming_edges.shape[0] ,\")\" \":\", incoming_edges.tolist())\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "cec1e59d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "153\n"
     ]
    }
   ],
   "source": [
    "node_ids = list(set(data_t[:,0].tolist()))\n",
    "print(len(node_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "716bad9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 14%|█▍        | 22/153 [01:58<11:46,  5.40s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[58], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mverify_heterodata_construction\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmonarch\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43medges_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnode_ids\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSuccesfully verified \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(node_ids)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m genes in the network!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[56], line 16\u001b[0m, in \u001b[0;36mverify_heterodata_construction\u001b[0;34m(data, edges_ddf, node_ids)\u001b[0m\n\u001b[1;32m     13\u001b[0m graph_out_count \u001b[38;5;241m=\u001b[39m dest[src \u001b[38;5;241m==\u001b[39m node_idx]\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     15\u001b[0m orig_in_count \u001b[38;5;241m=\u001b[39m edges_df[(edges_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mobject_id\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39mnode_idx) \u001b[38;5;241m&\u001b[39m (edges_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpredicate\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39medge_type_to_chk)][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mid\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mcompute()\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m---> 16\u001b[0m orig_out_count \u001b[38;5;241m=\u001b[39m \u001b[43medges_df\u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[43medges_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msubject_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43mnode_idx\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m&\u001b[39;49m\u001b[43m \u001b[49m\u001b[43m(\u001b[49m\u001b[43medges_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpredicate\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43medge_type_to_chk\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mid\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m0\u001b[39m]\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     20\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m graph_in_count \u001b[38;5;241m==\u001b[39m orig_in_count\n",
      "File \u001b[0;32m~/projects/def-mtarailo/rahit/python_environments/py311/lib/python3.11/site-packages/dask_expr/_collection.py:453\u001b[0m, in \u001b[0;36mFrameBase.compute\u001b[0;34m(self, fuse, **kwargs)\u001b[0m\n\u001b[1;32m    451\u001b[0m     out \u001b[38;5;241m=\u001b[39m out\u001b[38;5;241m.\u001b[39mrepartition(npartitions\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m    452\u001b[0m out \u001b[38;5;241m=\u001b[39m out\u001b[38;5;241m.\u001b[39moptimize(fuse\u001b[38;5;241m=\u001b[39mfuse)\n\u001b[0;32m--> 453\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDaskMethodsMixin\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/projects/def-mtarailo/rahit/python_environments/py311/lib/python3.11/site-packages/dask/base.py:375\u001b[0m, in \u001b[0;36mDaskMethodsMixin.compute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    351\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcompute\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[1;32m    352\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Compute this dask collection\u001b[39;00m\n\u001b[1;32m    353\u001b[0m \n\u001b[1;32m    354\u001b[0m \u001b[38;5;124;03m    This turns a lazy Dask collection into its in-memory equivalent.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    373\u001b[0m \u001b[38;5;124;03m    dask.compute\u001b[39;00m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 375\u001b[0m     (result,) \u001b[38;5;241m=\u001b[39m \u001b[43mcompute\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtraverse\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    376\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/projects/def-mtarailo/rahit/python_environments/py311/lib/python3.11/site-packages/dask/base.py:661\u001b[0m, in \u001b[0;36mcompute\u001b[0;34m(traverse, optimize_graph, scheduler, get, *args, **kwargs)\u001b[0m\n\u001b[1;32m    658\u001b[0m     postcomputes\u001b[38;5;241m.\u001b[39mappend(x\u001b[38;5;241m.\u001b[39m__dask_postcompute__())\n\u001b[1;32m    660\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m shorten_traceback():\n\u001b[0;32m--> 661\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mschedule\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdsk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    663\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m repack([f(r, \u001b[38;5;241m*\u001b[39ma) \u001b[38;5;28;01mfor\u001b[39;00m r, (f, a) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(results, postcomputes)])\n",
      "File \u001b[0;32m/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/python/3.11.5/lib/python3.11/queue.py:171\u001b[0m, in \u001b[0;36mQueue.get\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    170\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_qsize():\n\u001b[0;32m--> 171\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnot_empty\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m timeout \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m    173\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m must be a non-negative number\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m/cvmfs/soft.computecanada.ca/easybuild/software/2020/avx2/Core/python/3.11.5/lib/python3.11/threading.py:320\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    318\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[1;32m    319\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 320\u001b[0m         \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    321\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m    322\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "verify_heterodata_construction(monarch, edges_df, node_ids)\n",
    "print(f\"Succesfully verified {len(node_ids)} genes in the network!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4441a0fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from torch.utils.data import Dataset, DataLoader\n",
    "\n",
    "class ModifierDataset(Dataset):\n",
    "    def __init__(self, filepath: str = None):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            data (Tensor): A tensor containing node pairs and their similarity label.\n",
    "                           Shape: [num_pairs, 3], where each row is (node1, node2, label).\n",
    "        \"\"\"\n",
    "        self.data = torch.load(filepath)\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        modifier, target, label = self.data[idx]\n",
    "        return modifier, target, label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5966f8da",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = ModifierDataset(\"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/05_model_input/2024-03-31-merged-dataset.pt\")\n",
    "val_loader = DataLoader(dataset, batch_size=len(dataset), shuffle=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4dda1baf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(tensor(516790, dtype=torch.int32),\n",
       " tensor(518908, dtype=torch.int32),\n",
       " tensor(1, dtype=torch.int32))"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(iter(dataset))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "027f5d1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([     0,      1,  81868, ..., 558429, 558431, 558747], dtype=int32)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = torch.load(\"/home/rahit/projects/def-mtarailo/rahit/from_scratch/modspy-data/data/05_model_input/2024-03-31-merged-dataset.pt\")\n",
    "np.unique(data.numpy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f0f5514",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.5 (ModSpy CUDA) Kernel",
   "language": "python",
   "name": "modspy-py311-cuda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
